docs/design/README.md

```diff
@@ -0,0 +1,595 @@
+
+### Enhancing the Build System
+
+The existing build system will remain intact with certain features and abstractions added to it.
+Twoliter will serve as a facade providing a new interface over the existing build system.
+Twoliter will understand how to start a Bottlerocket build using the OOTB user's sources, packages etc.
+It will inject these directories into the existing build system by setting existing and new `Makefile.toml` variables.
+
+Twoliter will eliminate the need for all build-host software other than `docker`.
+A bootstrapping container will be published that contains any software needed to kick off and sustain a `cargo make` build command.
+
+## Kits
+
+A central idea in this design is that the build system will be able to produce and consume Docker images that contain RPM packages.
+We call these images *kits*.
+A kit is a Yum repo distributed in its entirety, atomically, as a container image.
+
+Maintainers may reference these kits in order to add packages to their Bottlerocket variants.
+In particular, the core project will publish all of its packages in a suite of kits.
+With these, an out-of-tree maintainer will have all the packages available that go into variants that the core project is currently producing.
+We envision maintainers other than the core team publishing Bottlerocket kits in the future.
+
+### Kit Signing
+
+In order to satisfy self-signing requirements, Kit RPMs will need to be GPG-signed.
+The kit will be distributed with the GPG public key, which means we need the container to also be signed.
+Existing container-signing mechanisms will be made available.
+(**TODO** - refine the design for signing artifacts, see https://github.com/bottlerocket-os/twoliter/issues/5)
+
+### Local and External Kits
+
+During the build process, there will be a distinction between *local kits* and *external kits*.
+A *local kit* is any kit where the packages are defined in the project tree and the maintainer wants to build them from source as part of the Bottlerocket build.
+An *external kit* is a kit that is used by pulling it from a docker registry, i.e. the container image has previously been built.
+
+It should not be a requirement for a maintainer to push a kit to a container registry.
+In other words, it should not be necessary to publish a kit in order to create a Bottlerocket variant.
+This is where *local kits* come in.
+Kits defined in the project tree need not be pre-built before a `twoliter build` command can succeed.
+Twoliter will understand that the local kits need to be built before the rest of the variant build can proceed.
+
+### Kits Project Directory
+
+Currently, there are three directories used for source code when building Bottlerocket:
+- `sources`: A Cargo workspace for first-party Rust code.
+- `packages`: A collection of Rust packages that build RPM packages.
+- `variants`: A Cargo workspace for defining what Bottlerocket variants.
+
+A new directory will be added: `kits`.
```

Comment:
Would kits be a replacement for some packages which are a pain for in-tree builds? xfsprogs comes to mind.

Reply:
> Would kits be a replacement for some packages which are a pain for in-tree builds? xfsprogs comes to mind.

That's sort of the idea. If you wanted to vend an `xfsprogs` package for multiple out-of-tree projects to use, you could create a kit container with it and any other packages it needs.

If you just need to use it locally in a single out-of-tree build project, you can sort of forget about "kits" since you won't be publishing any.

Either way the rpm spec for `xfsprogs` would exist in your out-of-tree twoliter project's `packages` directory.